### ISSUE TITLE
Feature suggestion: Analyzers

### ISSUE THREAD

**Issue by patriksvensson** [statiqdev/Statiq.Framework#104]
It would be useful if there existed a concept of analyzers in Statiq, where you could analyze the generated content and be able to generate warnings or errors. A use case for this might be to find documents missing specific properties in front-matter, or other scenarios where the content is correct from a technical standpoint, but might be incorrect in respect to external factors.

**Comment by daveaglick** [comment]
I _love_ this idea. I actually need something just like this for Discover .NET - that’s another use case: validating contributions in a PR build.

There’s an existing `ValidateMetadata<T>` module, but maybe this more general concept should replace/supplement that. There’s no reason a validation delegate shouldn’t be run on the document as a whole (so it can check content too). The idea of checking documents against other documents or external factors is also good.

I’d want to bring this into Statiq Web too in a way analyzers (validates?) are easy to define. Ideally we’d be able to add them via settings (like a configuration file) or via the bootstrapper.

Let me think on this a little - I’ll likely try to get it in a release soon(ish).

**Comment by daveaglick** [comment]
See https://github.com/terrajobst/sample-blog-tooling for another idea of how this could be done. In that repo, @terrajobst loads validators via reflection and then passes the MarkDig context for each Markdown document to the validator. I also _really_ like the idea of being able to classify validation results as either warnings or errors (diagnostics).

Another interesting aspect is how these validators inspect the MarkDig context and not the finished product. Something like that could be useful if I can figure out a generic way to insert analyzers at specific points in the process. Maybe it's as simple as defining an abstract base analyzer with specialized classes for analyzing output or whatever else (I.e. the base "Markdown" analyzer would be defined alongside the `RenderMarkdown` module and have a analyze overload the passes Markdown-specific information). An analyzer-aware module could ask for all analyzers that take a particular type of context object from the global list and then call them, not unlike how Statiq events are implemented. A more general `AnalyzeInputs<TAnalyzer>` module could also apply analyzers of the given type on the document inputs and could be placed anywhere. For example, in Statiq Web we'd probably place it at the end of each pipeline just before writing and maybe have a separate one with a different analyzer type for analyzing input files as they're read.

Also appears that Azure Pipelines and GitHub Actions support a special stdout syntax to capture these kinds of checks:
- https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#logissue-log-an-error-or-warning
- https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-a-warning-message

Ideally any output from analyzers/validators should use the correct GitHub Actions or Azure Pipelines output syntax to show in the log _if running on that platform_. It might also be helpful to integrate that into the logger generally (or make a new logger that gets added to the set of loggers in these environments)?

**Comment by patriksvensson** [comment]
@daveaglick It was actually something like that I had in mind when I wrote that 😀

**Comment by daveaglick** [comment]
Lol

![image](https://user-images.githubusercontent.com/1020407/93901429-8138e300-fcc4-11ea-86dc-dfdbbd3743d0.png)

**Comment by daveaglick** [comment]
Okay, think I've got a good design for this...

```csharp
public interface IValidator
{
    Task ValidateAsync(IValidationContext context);
    string[] Pipelines { get; }  // The pipeline(s) this validator applies to, null for all, empty for none
    Phase[] Phases { get; }  // The phase(s) this validator applies to, null for all, empty for none
}

public class ValidationResult
{
    public LogLevel { get; set; }
    public string Message { get; set; }
    public IDocument Document { get; set; }  // null if result doesn't apply to single document
}

public interface IValidationContext : IExecutionState
{
    // Various AddError(), AddWarning(), Add() methods

    ImmutableArray<IDocument> Documents { get; }
    string PipelineName { get; }
    IReadOnlyPipeline Pipeline { get; }
    Phase Phase { get; }
}

internal class ValidationContext : IValidationContext
{
    private readonly ConcurrentBag<ValidationResult> _results = ...;
}

// abstract base Validator, SyncValidator, DocumentValidator, and SyncDocumentValidator classes
```

- `IValidator` instances are added to a collection in the `Engine`
- The bootstrapper includes a method to find them via reflection (and it's used by default)
- Validators are called after every phase for every pipeline and filtered based on `Pipelines` and `Phases` properties
- Validators run in parallel
- Appropriate log message is generated for each `ValidationResult` and output _after_ all are run and results gathered
- Add a setting to treat all warning `ValidationResult` as errors

**Comment by daveaglick** [comment]
So now the hardest question..._Analyzer_, _Validator_, or something else? 🤔

**Comment by daveaglick** [comment]
I'm thinking validator is the right term. Analyzer makes sense in the context of Roslyn where things like code fixes can be suggested but aren't "wrong". In this case, the only reason for this to exist is to point out warnings or errors so it's more of a validation.

**Comment by daveaglick** [comment]
Hmm...this design doesn't really cover the situation where we want to validate some work product like the `MarkdownDocument` when rendering Markdown. That's a really useful use case for things like checking Markdown content.

On the other hand, I still like the general nature of this approach with well-known validation point (otherwise we get into tricky questions like "when do we validate?", "how can we avoid validating over and over?", etc.) My first thought about how to handle this generically is to use metadata to store the `MarkdownDocument` and then a special `MarkdownValidator` base class can be written that looks for that and passes it to a virtual validation method. Because the validators filter based on pipeline and phase, we'll also want to implement this special base inside Statiq Web so it can be correctly filtered for those pipelines. The downside of this approach is that it requires holding onto the `MarkdownDocument` instance inside document metadata which might impact memory utilization.

For that matter, a special `HtmlValidator` base could also be written that uses AngleSharp to create a DOM and passes that to the validation method for each document.

**Comment by daveaglick** [comment]
This has come together pretty well. Here's a Markdown validator based off of https://github.com/terrajobst/sample-blog-tooling/blob/master/src/blog-validate/Rules/VR06_FencedCodeBlocksMustHaveLanguage.cs from @terrajobst:

```csharp
public class FencedCodeBlocksMustHaveLanugage : SyncMarkdownValidator
{
    protected override void Validate(MarkdownDocument markdown, IDocument document, IValidationContext context)
    {
        IEnumerable<FencedCodeBlock> unfencedBlocks = markdown.Descendants<FencedCodeBlock>().Where(b => string.IsNullOrEmpty(b.Info));

        foreach (FencedCodeBlock block in unfencedBlocks)
        {
            context.AddError(document, $"Line {block.Line}: Fenced code blocks should specify a language");
        }
    }
}
```

And it produces:

![image](https://user-images.githubusercontent.com/1020407/94354910-9c517d00-004d-11eb-871f-d17df24d9af2.png)

![image](https://user-images.githubusercontent.com/1020407/94354936-bf7c2c80-004d-11eb-83de-239bea5f17e9.png)

Final step is to inject Azure Pipelines or GitHub Actions loggers that output additional console lines with the special CI syntax for warnings and errors so these show up.

**Comment by patriksvensson** [comment]
@daveaglick Would be nice if every validator has some kind of ID so they can be suppressed from the command line.

**Comment by daveaglick** [comment]
@patriksvensson 

> Would be nice if every validator has some kind of ID so they can be suppressed from the command line.

Agreed, and now that I've slept on it I've got some other changes in mind too. Sometimes you've got to implement something not quite right to get it perfect...

- I'm coming around on the name "analyzer" and think I made a mistake with "validator" (I know, I know, "Dave...a mistake?!"). I ended up thinking through a bunch of use cases where we wouldn't really be validating anything at all. One I particularly like is outputting a number of blog posts, next publish date, etc. - @khalidabuhakmeh has done something similar and it looks really useful.
- Not sure I love the analyzer specifying it's log level (`AddError()`, `AddWarning()`, etc.) - I'd rather have the analyzer say something and the user determine whether a particular analyzer produces errors, warnings, info, etc. The one challenge here is analyzer auto-registration which I'm currently doing - it scans the entry assembly for any analyzers and adds them automatically, so we'd need a way to say what level those should be. Not entirely sure how to address this yet - maybe a `DefaultAnalyzerLogLevelAttribute` or something.
- With regard to ID, I agree that'd be very helpful so you could register, change level, etc. from either the CLI or settings. An easy ID would be the name of the analyzer class. On the CLI I'm thinking `--analyzer [name] [level]` would either register or change the level. I.e. to remove one `--analyzer Foo None`.
- I would love something similar in settings too - an `Analyzers` key with a list of analyzers and their levels.

So...still some work to do but getting closer.

**Comment by khalidabuhakmeh** [comment]
My two-cents:

I've found that a strong tool set that complements writing has helped immensely.

If you folks want to chat about some nice-to-haves I'd be more than happy to.

**Comment by daveaglick** [comment]
@khalidabuhakmeh Absolutely - that's partly why I'm rethinking the design a bit, because I'm starting to think that we should deliver a growing set of analyzers that the user can added or remove at whatever log level they want. I.e. a default `CodeFencesShouldHaveLanguage` analyzer and I can either report on that by adding it with a `Information` level or make it a requirement by adding with an `Error` level.

Any use cases you have around analyzers that could assist the user would be awesome, whether they're of the validation variety or more "here's some helpful info".

**Comment by daveaglick** [comment]
Really happy with this now - analyzers can be added in multiple ways:
- All `IAnalyzer` implementations in the entry assembly are instantiated and added by default
- The Bootstrapper contains extensions to add analyzers by type, instance, or delegate
- The CLI can add analyzers and override log levels with `-a|--analyzer` options: `-a ValidateFooMetadata=Error`
- A `Analyzer` setting can also be used to add analyzers and override log levels (same syntax as the CLI)

Analyzers also have a name - by default the class name, but an alternate name can be specified (I.e. if using the same analyzer class with different parameters to do different things, you'll want to give each instance a different analyzer name).

This makes it very flexible, particularly for built-in analyzers. For example, I'm planning on shipping a variety of analyzers in Statiq Web. There will also be a Boostrapper `.AddDefaultAnalyzers()` (or similar) that registers them all with a warning or error level as appropriate. However, if you want a different level or want to register them individually without using the extension to register them all you can do so via the Bootstrapper, CLI, or `settings.yaml` file.

I'm going to consider this done in Framework. Statiq Web includes a `MarkdownAnalyzer` base class that passes a `MarkdownDocument` to the analyzer and I'm planning on adding an `HtmlAnalyzer` base that passes an AngleSharp DOM as well. Over time I'll add to the set of build-in analyzers. This will all ship in the next release (likely this week).
